ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.commons.ServerErrorCode;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.messageformat.MessageFormatFlags;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.GetOptions;
+import com.github.ambry.protocol.GetRequest;
+import com.github.ambry.protocol.PartitionRequestInfo;
+import com.github.ambry.utils.Time;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An abstract class for a get operation.
+ * @param <T> the type of the result of this operation.
+ */
+abstract class GetOperation<T> {
+  protected final RouterConfig routerConfig;
+  protected final ClusterMap clusterMap;
+  protected final ResponseHandler responseHandler;
+  protected final FutureResult<T> operationFuture;
+  protected final Callback<T> operationCallback;
+  protected final BlobId blobId;
+  protected final Time time;
+  protected boolean operationCompleted = false;
+  protected Exception operationException;
```

Comment:
Should this be specifically declared as `RouterException` or generally as `Exception`? Any situation for the latter?